### PR TITLE
[GOBBLIN-707] rewrite gobblin script to combine all modes and command

### PR DESCRIPTION
**Pull request by jhsenjaliya**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [*] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-707
    - https://issues.apache.org/jira/browse/GOBBLIN-694
    - https://issues.apache.org/jira/browse/GOBBLIN-477

### Description
- [*] Here are some details about my PR, including screenshots (if applicable):
This PR adds new gobblin.sh script that does following:
1. there were individual scripts for individual gobblin modes and gobblin commands, all does config, log and other param management differently, this PR combines all commands and modes into one script that not only standardize it but also makes easy for devs to use it. 
2. config were scattered or duplicated per gobblin mode, this PR standardizes
3. removes scripts and redundant configs.
4. removes usage of ENV variables for configurations (GOBBLIN_JOB_CONFIG_DIR, GOBBLIN_FWDIR, GOBBLIN_LOG_DIR, GOBBLIN_WORK_DIR, etc... ) 
5. updates the docs accordingly.

### Tests
- [*] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:  These are scripts and config changes, dont require test case.

### Commits
- [*] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: bin/gobblin.sh [hunk 1]
```diff
@@ -17,50 +17,410 @@
...
+        status
+    ;;
+    "restart")
+        stop
+        sleep 2
+        start
+    ;;
+    *)
+        print_usage
+#        if [[ "$GOBBLIN_MODE" == "$CLI_MODE" ]] || [[ "$GOBBLIN_MODE" == "$STATESTORE_CHECK_MODE" ]] || [[ "$GOBBLIN_MODE" == "$STATESTORE_CLEAN_MODE" ]]; then
```
Comment (jhsenjaliya) [inline bin/gobblin.sh]: I will remove this commented code.

File: gobblin-docs/user-guide/Gobblin-CLI.md [hunk 2]
```diff
@@ -28,29 +28,29 @@ Gobblin ingestion applications
 
 Gobblin ingestion applications can be accessed through the command `run`:
 ```bash
-bin/gobblin run [listQuickApps] [<quick-app>] -jobName <jobName> [OPTIONS]
+bin/gobblin cli run [listQuickApps] [<quick-app>] -jobName <jobName> [OPTIONS]
```
[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: @autumnust, with this change, I am targetting to combine all services and commands together because the config and other mgmt is the same for all services and commands. since the gobblin is just link to gobblin.sh, with this change it will handle the cli as a command. so following becomes the new way of using gobblin script for various activities.
`gobblin.sh  <command> <params>
gobblin.sh  <service-name> <start|stop|status>

commands values: admin, cli, statestore-check, statestore-clean, historystore-manager
service values: standalone, cluster-master, cluster-worker, aws, yarn, mr, service`

[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: @autumnust , does that make sense ?

File: conf/yarn/reference.conf [hunk 3]
```diff
@@ -38,6 +38,6 @@ gobblin.yarn.work.dir=/gobblin
 gobblin.cluster.helix.cluster.name=GobblinYarn
 gobblin.cluster.zk.connection.string="localhost:2181"
 
-fs.uri="hdfs://localhost:9000"
```
Comment (autumnust) [inline conf/yarn/reference.conf]: Just curious, why port number needs to be changed here? Seems both 8020/9000 can be default port number for IPC of Namenode

File: conf/yarn/application.conf [hunk 4]
```diff
@@ -22,15 +22,18 @@ gobblin.yarn.app.name=GobblinYarn
 gobblin.yarn.app.master.memory.mbs=256
 gobblin.yarn.initial.containers=2
 gobblin.yarn.container.memory.mbs=512
-gobblin.yarn.conf.dir=<directory where Gobblin on Yarn related configuration files are located>
-gobblin.yarn.lib.jars.dir=<directory where Gobblin on Yarn lib jars are located>
-gobblin.yarn.app.master.files.local=${gobblin.yarn.conf.dir}"/log4j-yarn.properties,"${gobblin.yarn.conf.dir}"/application.conf,"${gobblin.yarn.conf.dir}"/reference.conf"
+gobblin.yarn.conf.dir=/tools/gobblin-dist/conf/yarn/
```
Comment (autumnust) [inline conf/yarn/application.conf]: Why this is being hard coded ?

File: gobblin-docs/user-guide/Gobblin-CLI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (autumnust) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: Sorry for getting back to this late. 
Yes that makes sense to me. I see them in `gobblin.sh` printUsage() method. Can you add/edit documentation to mention the existence of `gobblin.sh` so that newer users will be aware of it ?

File: conf/yarn/reference.conf (hunk 3 above)
Comment (jhsenjaliya) [reply conf/yarn/reference.conf]: yes, both are acceptable, just changing it to 8020 as default which i believe most people use, can change it to 9000 if its otherwise, no prob.

File: conf/yarn/application.conf (hunk 4 above)
Comment (jhsenjaliya) [reply conf/yarn/application.conf]: this is missed, let me change this to `gobblin.yarn.conf.dir=${GOBBLIN_HOME}/conf/yarn/` thanks for catching this, this was my local config.

File: gobblin-docs/user-guide/Gobblin-CLI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: ok sure, it requires log of doc changes, and some reorganization, which i can take care of but can we get #2586 merged? otherwise i ll have lot of conflicts.

**Comment by jhsenjaliya** [comment]
Hi @autumnust, working on adding and refactoring documentation for commands and deployments.... once #2586 is merged, I will add the commit here.

**Comment by jhsenjaliya** [comment]
@autumnust , updated and refactored the doc, please take a look.

**Comment by autumnust** [comment]
@jhsenjaliya  There's seems to be an travis issue but I am not sure if it is related to your changes.

**Comment by jhsenjaliya** [comment]
@autumnust , Thanks for checking on that, I will push some changes to re-trigger it after some review.

File: gobblin-docs/user-guide/Gobblin-CLI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibuenros) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: @jhsenjaliya Gobblin CLI is already an aggregator for functionality, no reason to add an additional `cli` to the command. Today there is `bin/gobblin run` which should stay as is, additionally there can be `bin/gobblin admin`, etc.

[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: Hi @ibuenros ,  the `cli` has lot of options other than `run`, like `passwordManager`, `watermarks`, etc...  so we can keep it under `cli` category for now, and in future, those options can be added directly to the `gobblin.sh` if we really want to have `run` command directly under gobblin  without `cli`.
here the problem I m trying to solve is mostly inconsistency and lack of clarity among functionality of all gobblin scripts. I have also added more detail GOBBLIN-707 jira ticket to explain why we should combine all these scripts.

[the remark this answers is not in the document]
Comment (ibuenros) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: My point is that there is no reason to move all of those applications down one level. Instead, everything else that you would want to add to the `bin/gobblin` command should be implemented as a `CliApplication`.

I do see the benefit of aggregating all functionality, I just think it should be done without changing current commands.

[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: ok , sure.
is `gobblin admin <args>` fine ?
is it only about `gobblin cli run <args>`? then i can bring the `run`, `watermarks`, `passwrodManager`, etc commands to gobblin.sh level so that the command signature wont change from existing one.

File: bin/gobblin.sh [hunk 5]
```diff
@@ -17,50 +17,435 @@
...
-  echo $CLASSPATH
+# JAVA_HOME is required.
+if [[ -z "$JAVA_HOME" ]]; then
+    echo -e "\nError: Environment variable JAVA_HOME not set!\n"
+    exit 1
+fi
+
+# global vars
+
+GOBBLIN_VERSION="0.15.0"
```
Comment (jhsenjaliya) [inline bin/gobblin.sh]: will replace this with @project.version@ to get it placed by gradle.

File: gobblin-docs/user-guide/Gobblin-CLI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibuenros) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: What is the proposal on how to do that though? Actually I'm a bit confused with this PR, there are a lot of removed shell scripts, but there seems to be no replacement.

[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: to handle it on bash script level and call the `GobblinCli` with appropriate args for each command accordingly.
btw, each removed shell script has support to execute via this updated `gobblin.sh`.
if you can pull this PR and try some of the commands, it will be clear on how things are working with these updates. git diff is bit confusing for such large PR. Thanks

[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: @ibuenros , lets take this offline over email. Thanks

[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply gobblin-docs/user-guide/Gobblin-CLI.md]: Hi @ibuenros, I have added the mapping of old gobblin script to new gobblin command in jira ticket and also started email thread: "GOBBLIN-707 Review", would you please take a look and reply there. Thanks

**Comment by jhsenjaliya** [comment]
@autumnust , updated docs and also added new info in doc regarding the usage of gobblin.sh, please take a look when you get chance. Thanks

**Comment by jhsenjaliya** [comment]
@ibuenros , @autumnust, Would you pls take a look ? Thanks

**Comment by jhsenjaliya** [comment]
1 out of 4 test env has failed, will look into it.

File: bin/gobblin.sh [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
Comment (ibuenros) [inline bin/gobblin.sh]: Can we leave `gobblin.sh` relatively simple and instead have `gobblin-cli.sh` and `gobblin-service.sh`? `gobblin.sh` would just redirect to the correct place depending on the first argument.

**Comment by jhsenjaliya** [comment]
@ibuenros, @autumnust  since our last discussion, I have also brought all CLI and command classes under `GobblinCli` with Alias, so its much organized now that also helps clear CLI options.
I would suggest try to check out this PR locally and play around with it to see how this works.
Thanks

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixRetriggeringJobCallable.java [hunk 7]
```diff
@@ -152,7 +152,7 @@ public Void call() throws JobException {
         GobblinClusterConfigurationKeys.JOB_ALWAYS_DELETE,
         "false");
 
-    try {
+    try { //TODO: what is really the difference ?
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixRetriggeringJobCallable.java]: Why is this relevant to this PR?

File: bin/gobblin.sh (hunk 6 above)
Comment (sv2000) [reply bin/gobblin.sh]: @jhsenjaliya since this is a big change, and the fact that we use the existing shell scripts in some of our internal use cases, it would be great if you can provide documentation on how to run common commands using the new script.

File: bin/gobblin-admin.sh [hunk 8]
```diff
@@ -1,142 +0,0 @@
-#!/bin/bash
```
Comment (htran1) [inline bin/gobblin-admin.sh]: How about maintaining compatibility with callers of the old script? One option is to leave the old scripts as wrapper scripts that call the new one. Another option is to have the old script names be symlinks to the new script and detect the usage based on the value of $0.

File: bin/gobblin.sh (hunk 6 above)
Comment (jhsenjaliya) [reply bin/gobblin.sh]: yes, i have updated the doc with updated help and sample commands.

File: bin/gobblin.sh [hunk 9]
```diff
@@ -17,50 +17,488 @@
...
+
+function print_gobblin_usage() {
+    echo "Usage:"
+    echo "gobblin.sh  cli     <cli-command>    <params>"
+    echo "gobblin.sh  service <execution-mode> <start|stop|status>"
+    echo ""
+    echo "Use \"gobblin <cli|service> --help\" for more information.         (Gobblin Version: $GOBBLIN_VERSION)"
+}
+
+function print_gobblin_cli_usage() {
```
[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply bin/gobblin.sh]: Yes, GobblinCLI has `printusage` which will list all the Alias classes, but it wont display other help options. Also its better to have all help options at script level rather than at class level. 
btw both are available depending on the usage.
```
bin/gobblin cli somecommand
Could not find an application with alias somecommand
Available commands:
	job-state-to-json		To convert Job state to JSON
	jobs		Command line job info and operations
	passwordManager		Encrypt or decrypt strings for the password manager.
	run		Run a Gobblin application.
	decrypt		Decryption utilities
	job-store-schema-manager		Database job history store schema manager
	stateMigration		Command line tools for migrating state store
	keystore		Examine JCE Keystore files
	config		Query the config library
	watermarks		Inspect streaming watermarks
	cleaner		Data retention utility```

File: conf/cluster-master/application.conf [hunk 10]
```diff
@@ -69,3 +70,20 @@ task.status.reportintervalinms=1000
 # Enable metrics / events
 metrics.enabled=true
 
+# UI
+admin.server.enabled=true
```
[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply conf/cluster-master/application.conf]: so that first time user dont have to figure this out, if there is any specific reason not to turn it on by default, then i can remove this config. I have added those configs that I had to add when i setup the gobblin first time to get it working.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixRetriggeringJobCallable.java (hunk 7 above)
Comment (jhsenjaliya) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixRetriggeringJobCallable.java]: oh this got into this by mistake, it was my own code comment. will remove it.

File: bin/gobblin-admin.sh (hunk 8 above)
Comment (jhsenjaliya) [reply bin/gobblin-admin.sh]: I was targeting to cleanup and simplify those too many scripts Gobblin has, which could be confusing for new users. user should really move to new scripts, so not sure how much the backward compatibility is required for scripts, as oppose to APIs, otherwise we will never be able to clean up. but I see ur point so may be I will create wrapper as you suggested with note to remove that in future.

**Comment by jhsenjaliya** [comment]
@htran1, @sv2000 , added wrapper scripts, pls take a look.

File: conf/cluster-master/application.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (jhsenjaliya) [reply conf/cluster-master/application.conf]: @ibuenros, made default to false as before, since it requires store config as well.

**Comment by jhsenjaliya** [comment]
@sv2000 @htran1 @ibuenros , added some more changes as per my usage. pls take a look.

**Comment by jhsenjaliya** [comment]
@sv2000 @htran1 @ibuenros, can you guys pls take a look at this ?

**Comment by sv2000** [comment]
@jhsenjaliya looks like there are conflicts. Can you take a look? Once you resolve the conflicts, I can merge the PR.

**Comment by jhsenjaliya** [comment]
@sv2000, cant figure out what is the conflict here, i can not see the diff, since i dont have write access. can u pls take a look?  btw `conf/service` is renamed to `conf/gobblin-as-service`.

**Comment by jhsenjaliya** [comment]
End up fixing the conflict with merge. Please take a look. Thanks
